### PR TITLE
interfaces/pwm: add PWM interface

### PR DESCRIPTION
**Pull request by woodrow-shen**
PWM interface is often used to control device's LED, fan or heater and
its sysfs can allow user space to expose PWM under /sys/class/pwm/pwmchipN
where N is the base of PWM chip and each chip may have multiple channels
used for application. The access path for the specific channel should be
specified at /sys/class/pwm/pwmchipN/pwm? and that channel needs to be
exported via /sys/class/pwm/pwmchipN/export, so its behavior is similar
to GPIO interface.

Reference:
https://www.kernel.org/doc/Documentation/pwm.txt

### CODE REVIEW

File: interfaces/builtin/pwm.go
Comment (anonymouse64) [inline interfaces/builtin/pwm.go:36]: this matches the gpio declaration and looks correct to me

File: interfaces/builtin/pwm.go
Comment (anonymouse64) [inline interfaces/builtin/pwm.go:65]: these things can all be more easily specified by embedded the `commonInterface` inside the `pwmInterface` like was done for the docker interface: 

https://github.com/snapcore/snapd/blob/5831ee159bba3c432dab43875206355089678281/interfaces/builtin/docker_support.go#L651-L653

and

https://github.com/snapcore/snapd/blob/5831ee159bba3c432dab43875206355089678281/interfaces/builtin/docker_support.go#L706-L719

File: interfaces/builtin/pwm.go
Comment (anonymouse64) [inline interfaces/builtin/pwm.go:141]: more readable versions of these would be:

```suggestion
		ExecStart:       fmt.Sprintf("/bin/sh -c 'test -e /sys/class/pwm/pwmchip%[1]d/pwm%[2]d || echo %[2]d > /sys/class/pwm/pwmchip%[1]d/export'", chipNum, pwmNum),
		ExecStop:        fmt.Sprintf("/bin/sh -c 'test ! -e /sys/class/pwm/pwmchip%[1]d/pwm%[2]d || echo %[2]d > /sys/class/pwm/pwmchip%[1]d/unexport'", chipNum, pwmNum),
```

though the same comment applies to the existing gpio interface too

File: interfaces/builtin/pwm_test.go
Comment (anonymouse64) [inline interfaces/builtin/pwm_test.go:147]:
```suggestion
	// pwm slot on OS accepted
```

**Review by anonymouse64 — COMMENTED** [review]
some little nitpicky things, but overall LGTM. We still need to get a security review for this though and the naming should be approved by @pedronis but otherwise I think this is very close.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:36]: Yeap I just follow up the code from gpio and changed the summary a bit.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:65]: Thanks for pointing it out. Let me update this accordingly.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:141]: Sure thing!

Comment (woodrow-shen) [reply interfaces/builtin/pwm.go]: Updated as well.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go]: Applied commonInterface for PWM now! Please let me if something wrong.

File: interfaces/builtin/pwm_test.go
Comment (anonymouse64) [inline interfaces/builtin/pwm_test.go:155]: I think we actually need another test case here for the apparmor snippets:

```go
func (s *PwmInterfaceSuite) TestApparmorConnectedSlot(c *C) {
	spec := &apparmor.Specification{}
	err := spec.AddConnectedPlug(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
	c.Assert(err, IsNil)
	c.Assert(spec.Snippets(), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
}
```

**Review by anonymouse64 — APPROVED** [review]
lgtm, but we need one more test I think after re-reviewing the PR, thanks for all the work on this

File: interfaces/builtin/pwm.go
Comment (alexmurray) [inline interfaces/builtin/pwm.go:54]: typo - s/chanel/channel/

File: interfaces/builtin/pwm.go
Comment (alexmurray) [inline interfaces/builtin/pwm.go:106]: I wonder what your thoughts would be on having explicit entries for the various files under `pwm%d` with the appropriate permissions - ie `period`, `duty_cycle`, `polarity` and `enable` rather than just `*` - otherwise if in a future kernel update there are new files added here that grant extra permissions this will be implicitly granted access as well?

File: interfaces/builtin/pwm_test.go
Comment (alexmurray) [reply interfaces/builtin/pwm_test.go:155]: Yes please - thanks for raising this @anonymouse64

**Review by alexmurray — CHANGES_REQUESTED** [review]
Overall this seems reasonable and is comparable to the existing `gpio` interface, however can you please add the test for the AppArmorConnectedSlot snippet and have a think about whether it is perhaps better to have the various files exported by the sysfs interface explicitly listed rather than just `*` for a more defensive approach to any possible future changes to this interface within the kernel. Thanks.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:54]: noted

File: interfaces/builtin/pwm_test.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm_test.go:155]: @anonymouse64 Thanks for hints! I tried to apply your suggestion but I always got the following:
```
FAIL: pwm_test.go:155: PwmInterfaceSuite.TestApparmorConnectedSlot

pwm_test.go:159:
    c.Assert(spec.Snippets(), testutil.Contains, "/sys/class/pwm/pwmchip10/pwm100/* rwk")
... container map[string][]string = map[string][]string{}
... elem string = "/sys/class/pwm/pwmchip10/pwm100/* rwk"
... container has items of type []string but expected element is a string

OOPS: 1203 passed, 1 FAILED
--- FAIL: Test (0.94s)
```
so how can I fix this as I'm not familiar with test code actually.

**Comment by pedronis** [comment]
added myself to check on the name

File: interfaces/builtin/pwm.go
Comment (pedronis) [reply interfaces/builtin/pwm.go:106]: fwiw gpio has also /* there

File: interfaces/builtin/pwm.go
Comment (pedronis) [inline interfaces/builtin/pwm.go:55]: is this number the:

per-chip relative index of the PWM device

or the 

global index of the PWM device

(I'm looking at https://www.kernel.org/doc/html/latest/driver-api/miscellaneous.html#c.pwm_device)

**Comment by pedronis** [comment]
the name of the interface is fine.  I wonder if "number" is the clearest name for the attribute though, I asked a related question in one of the comments.

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:55]: @pedronis It's per-chip index to channel exactly. For example, I take the test code to use:

```
slots:
    my-pin:
        interface: pwm
        chip-number: 10
        number: 100
```
so that required index (aka number) of chip10 under sysfs will be `/sys/class/pwm/pwmchip10/pwm100/`.

Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:55]: @pedronis It's per-chip index to channel exactly. For example, I take the test code to use:

```
slots:
    my-pin:
        interface: pwm
        chip-number: 10
        number: 100
```
so that required index (aka number here) of chip10 under sysfs will be `/sys/class/pwm/pwmchip10/pwm100/`.

From above description, `pwm_chip` provides linked-list to point `pwm_device` which exposed one and more PWM signal sources, so if I understand correctly the relationship between PWM interface and document should be:

per-chip relative index of the PWM device -> number
global index of the PWM device -> chip-number

Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:55]: Maybe given an idea roughly:

number -> perchip-index
chip-number -> global-index

Comment (anonymouse64) [reply interfaces/builtin/pwm.go:55]: what about an attribute of `channel` instead? so then we have:
```
slots:
  my-pin:
    interface: pwm
    chip-number: 10
    channel: 100
```
?

File: interfaces/builtin/pwm_test.go
Comment (anonymouse64) [inline interfaces/builtin/pwm_test.go:88]: so in order to test the apparmor snippets I mention below, you will need to add an app to the gadget, though a hook would work too and is perhaps more realistic, so something like:
```suggestion
    bad-interface-plug: other-interface
apps:
    svc:
        command: bin/foo.sh
```

then the snippet I recommended about the AppArmor snippet test would become:

```go
func (s *PwmInterfaceSuite) TestApparmorConnectedSlot(c *C) {
	spec := &apparmor.Specification{}
	err := spec.AddConnectedPlug(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
	c.Assert(err, IsNil)
	c.Assert(spec.SnippetForTag("snap.my-device.svc"), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
}
```

Comment (anonymouse64) [reply interfaces/builtin/pwm_test.go:88]: @woodrow-shen ^

File: interfaces/builtin/pwm.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm.go:55]: I'm ok with above +1, @pedronis, can we finalize that?

File: interfaces/builtin/pwm_test.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm_test.go:88]: Still go the error like
```
pwm_test.go:164:
    c.Assert(spec.SnippetForTag("snap.my-device.svc"), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
... container string = "" 
... elem string = "/sys/class/pwm/pwmchip10/pwm100/* rwk" 

FAIL: pwm_test.go:160: PwmInterfaceSuite.TestApparmorConnectedSlot
```

Btw, I suspect that the function name should be `TestApparmorConnectedPlug` instead of `TestApparmorConnectedSlot`? Coz I think `TestApparmorConnectedPlug` in the test code could map to `AppArmorConnectedPlug` to make sense?

And, I also found that gpio_test.go didn't have `TestApparmorConnectedPlug` as it also implemented `AppArmorConnectedPlug`, but maybe some stories behind the history...

Comment (anonymouse64) [reply interfaces/builtin/pwm_test.go:88]: ok, sorry I forgot you need to mock the evalSymlink function in order for the policy to generate, something like this should work for you: https://pastebin.ubuntu.com/p/btsYXQGrGY/

Comment (woodrow-shen) [reply interfaces/builtin/pwm_test.go:92]: Thanks @anonymouse64 helping a lot on the testing! Just updated accordingly. TIL!

File: interfaces/builtin/pwm_test.go
Comment (anonymouse64) [inline interfaces/builtin/pwm_test.go:196]: as was pointed out in my PR, this should actually be called:

```suggestion
func (s *PwmInterfaceSuite) TestApparmorConnectedPlug(c *C) {
```

but don't worry about updating it here, we can do it in a followup or I can push a commit here if tests are red due to unrelated failures and would need to be retriggered anyways

**Review by anonymouse64 — APPROVED** [review]
LGTM, thanks for the additional test, the attribute naming question is the only remaining thing for this PR

File: interfaces/builtin/pwm_test.go
Comment (woodrow-shen) [reply interfaces/builtin/pwm_test.go]: Let's get it updated!

File: interfaces/builtin/pwm.go
Comment (pedronis) [reply interfaces/builtin/pwm.go:55]: channel is fine with me (it has a meaning for snaps but given the context is should be fine)

Comment (woodrow-shen) [reply interfaces/builtin/pwm.go]: Done for change from `number` to `channel`. Thx.

**Review by pedronis — APPROVED** [review]
lgtm

File: interfaces/builtin/pwm.go
Comment (alexmurray) [reply interfaces/builtin/pwm.go]: Yes, I saw that precedent, but I thought perhaps it might be better to take a more defensive approach - and so I wondered what others opinions on this would be? I prefer to be more specific with these interfaces but if the general approach is to be less specific and more flexible (ie `*` in this case) that is fine, I just want it to be a conscious choice and to make sure we take a consistent approach when adding new interfaces like this in the future.

Comment (anonymouse64) [reply interfaces/builtin/pwm.go]: Personally, I think for an interface like this which already grants write access to low level devices and doesn't auto-connect, I think it's okay to be more permissive with `*`, but if it were a read-only interface or something that auto-connected I think we should be more specific, but I will defer to @pedronis on the general direction we want to take going forward on these interfaces

Comment (pedronis) [reply interfaces/builtin/pwm.go]: that access at the end is scoped down to a pin/channel so unless the kernel makes very unorthodox choices the relevant surface should be contained

Comment (mvo5) [reply interfaces/builtin/pwm.go]: I agree with Ian here and think this is okay. I will merge given that it seems this is not super controversial, then we can always tweak/improve before this lands on master.

**Comment by mvo5** [comment]
The failure:
```
2021-04-07T13:52:29.9217336Z ##[error]2021-04-07 13:52:29 Error executing google:ubuntu-16.04-64:tests/main/interfaces-many-core-provided (apr071339-039127) : 
...
2021-04-07T13:52:30.4011779Z + snap connect test-snapd-policy-app-consumer:pwm :pwm
2021-04-07T13:52:30.4012739Z error: cannot perform the following tasks:
2021-04-07T13:52:30.4014595Z - Connect test-snapd-policy-app-consumer:pwm to core:pwm (cannot obtain systemd services for snap "core": snap "core" does not have attribute "chip-number" for interface "pwm")
```
looks real

**Comment by anonymouse64** [comment]
I pushed a fix for the failing spread test here

**Comment by alexmurray** [comment]
If it wasn't clear, +1 from me on this.

**Comment by woodrow-shen** [comment]
Thanks everyone for supporting this feature :)

**Pull request merged**
